### PR TITLE
[intfsorch] add RIF flex counter group

### PR DESCRIPTION
**Pull request by mykolaf**
Signed-off-by: Mykola Faryma <mykolaf@mellanox.com>

**What I did**
 - add RIF counter flex group
 - add logic to register RIF objects in Flex Counter
**Why I did it**

**How I verified it**

**Details if related**

### CODE REVIEW

File: orchagent/intfsorch.cpp [hunk 1]
```diff
@@ -652,3 +697,112 @@ void IntfsOrch::removeDirectedBroadcast(const Port &port, const IpAddress &ip_ad
...
+            addRifToFlexCounter(id, it->m_alias, type);
+            it = m_rifsToAdd.erase(it);
+        }
+        else
+        {
+            ++it;
+        }
+    }
+    
+    // for (auto it = m_rifsToRemove.begin(); it != m_rifsToRemove.end();)
```
Comment (marian-pritsak) [inline orchagent/intfsorch.cpp]: Commented out by mistake or irrelevant?

**Comment by mykolaf** [comment]
depend on schema Azure/sonic-swss-common#256

**Comment by mykolaf** [comment]
retest this please

**Comment by lguohan** [comment]
retest this please

File: orchagent/intfsorch.cpp [hunk 2]
```diff
@@ -30,10 +31,49 @@ extern BufferOrch *gBufferOrch;
...
+
 IntfsOrch::IntfsOrch(DBConnector *db, string tableName, VRFOrch *vrf_orch) :
         Orch(db, tableName, intfsorch_pri), m_vrfOrch(vrf_orch)
 {
     SWSS_LOG_ENTER();
+
+    /* Initialize DB connectors */ 
+    m_counter_db = shared_ptr<DBConnector>(new DBConnector(COUNTERS_DB, DBConnector::DEFAULT_UNIXSOCKET, 0));
+    m_flex_db = shared_ptr<DBConnector>(new DBConnector(FLEX_COUNTER_DB, DBConnector::DEFAULT_UNIXSOCKET, 0));
+    // m_state_db = shared_ptr<DBConnector>(new DBConnector(STATE_DB, DBConnector::DEFAULT_UNIXSOCKET, 0));
```
Comment (prsunny) [inline orchagent/intfsorch.cpp]: Please remove commented code

File: orchagent/intfsorch.cpp [hunk 3]
```diff
@@ -671,3 +715,104 @@ void IntfsOrch::removeDirectedBroadcast(const Port &port, const IpAddress &ip_ad
...
+
+void IntfsOrch::doTask(SelectableTimer &timer)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_NOTICE("Registering %ld new intfs, deleting %ld ", m_rifsToAdd.size(), m_rifsToRemove.size());
+    string value;
+    for (auto it = m_rifsToAdd.begin(); it != m_rifsToAdd.end(); ++it)
+    {
+        SWSS_LOG_NOTICE("Registering begin -> %s ", it->m_alias.c_str());
```
Comment (prsunny) [inline orchagent/intfsorch.cpp]: Why do we have a separate for-loop for this log? With the log in below for-loop, this appears to be redundant.

File: orchagent/intfsorch.cpp [hunk 4]
```diff
@@ -671,3 +715,104 @@ void IntfsOrch::removeDirectedBroadcast(const Port &port, const IpAddress &ip_ad
...
+    }
+    for (auto it = m_rifsToAdd.begin(); it != m_rifsToAdd.end(); )
+    {
+        const auto id = sai_serialize_object_id(it->m_rif_id);
+        SWSS_LOG_NOTICE("Registering %s, id %s", it->m_alias.c_str(), id.c_str());
```
Comment (prsunny) [inline orchagent/intfsorch.cpp]: Suggest to change this to INFO level

File: orchagent/intfsorch.cpp [hunk 5]
```diff
@@ -671,3 +715,104 @@ void IntfsOrch::removeDirectedBroadcast(const Port &port, const IpAddress &ip_ad
...
+                type = "SAI_ROUTER_INTERFACE_TYPE_VLAN";
+                break;
+            default:
+                SWSS_LOG_ERROR("Unsupported port type: %d", it->m_type);
+                type = "";
+                break;
+        }
+        if (m_vidToRidTable->hget("", id, value))
+        {
+            SWSS_LOG_NOTICE("Registering %s it is ready", it->m_alias.c_str());
```
Comment (prsunny) [inline orchagent/intfsorch.cpp]: Suggest INFO level

File: orchagent/intfsorch.cpp [hunk 6]
```diff
@@ -671,3 +715,104 @@ void IntfsOrch::removeDirectedBroadcast(const Port &port, const IpAddress &ip_ad
...
+void IntfsOrch::generateInterfaceMap()
+{
+    m_updateMapsTimer->start();
+}
+
+void IntfsOrch::doTask(SelectableTimer &timer)
+{
+    SWSS_LOG_ENTER();
+
+    SWSS_LOG_NOTICE("Registering %ld new intfs, deleting %ld ", m_rifsToAdd.size(), m_rifsToRemove.size());
```
Comment (prsunny) [inline orchagent/intfsorch.cpp]: Suggest to change to INFO/DEBUG level

File: orchagent/intfsorch.h [hunk 7]
```diff
@@ -35,11 +38,33 @@ class IntfsOrch : public Orch
...
+    void generateInterfaceMap();
+    void addRifToFlexCounter(const string&, const string&, const string&);
+    void removeRifFromFlexCounter(const string&, const string&);
+
     bool setIntf(const string& alias, sai_object_id_t vrf_id = gVirtualRouterId, const IpPrefix *ip_prefix = nullptr);
 private:
+
+    SelectableTimer* m_updateMapsTimer = nullptr;
+    std::vector<Port> m_rifsToAdd;
+    std::vector<Port> m_rifsToRemove;
```
Comment (prsunny) [inline orchagent/intfsorch.h]: I don't see this updated any where. Is it missed?

File: orchagent/intfsorch.cpp [hunk 8]
```diff
@@ -466,6 +507,9 @@ bool IntfsOrch::removeRouterIntfs(Port &port)
         return false;
     }
 
+    const auto id = sai_serialize_object_id(port.m_rif_id);
+    removeRifFromFlexCounter(id, port.m_alias);
```
Comment (prsunny) [inline orchagent/intfsorch.cpp]: Why is this called directly whereas `addRifToFlexCounter `is based on timer?

**Comment by mykolaf** [comment]
retest this please

**Comment by mykolaf** [comment]
retest this please

**Comment by mykolaf** [comment]
retest this please

**Comment by mykolaf** [comment]
retest this please

File: orchagent/intfsorch.cpp (hunk 8 above)
Comment (mykolaf) [reply orchagent/intfsorch.cpp]: this is due to FC logic.
The Flex counter will try to determine the object's type based on the vid. The VIDTOIRID is not instantly available on create.
For the delete flow it's  the opposite.

If there will be some issues with this logic, I believe the FC flow should be changed. Right now it's tricky to dynamically add/remove objects that need to be polled.

Comment (tieguoevan) [reply orchagent/intfsorch.cpp]: How to ensure deleting flex counter before deleting rif in syncd? If deleting rif happened before deleting flex counter, rid cannot be found in the VIDTOIRID.

**Pull request merged**